### PR TITLE
Add unit tests to watcher + move ignored extensions

### PR DESCRIPTION
**Pull request by DanielSz50**

### CODE REVIEW

File: internal/watcher/watcher_test.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	defer os.RemoveAll(baseDir)
+
+	subDir, err := ioutil.TempDir(baseDir, "watcher_test2")
+	if err != nil {
+		t.Fatalf("ioutil.TempDir: %v", err)
+	}
+
+	for i := 0; i < 10; i++ {
+		file := filepath.Join(baseDir, fmt.Sprintf("test_file_%d", i))
+		if err = ioutil.WriteFile(file, []byte(":octopus:"), 0600); err != nil {
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:33]: I suggest using the default operating system's permission for the file creation.

File: internal/watcher/watcher_test.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+					assert.Fail(t, "Haven't excluded files with ignored extensions.")
+					return
+				}
+			}
+		}
+	})
+}
+
+type mockWatchFuncProvider struct {
+	hasBeenCalled bool
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:76]: `called` would be sufficient, I think.

File: internal/watcher/watcher_test.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+	files, directories, err := GetAllInDirectory(baseDir, []string{".swp"})
+	if err != nil {
+		t.Fatalf("GetAllInDirectory: %v", err)
+	}
+
+	done := make(chan struct{})
+
+	go func() {
+		time.Sleep(1 * time.Second)
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:107]: It's a tiny bit non-deterministic right not. How about utilizing [parallel testing](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks)?

File: internal/watcher/watcher_test.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			token:            "",
+		}
+
+		actual, err := NewDifferential([]string{}, nil, "", "")
+		if err != nil {
+			t.Fatalf("NewDifferential: %v", err)
+		}
+
+		assert.Equal(t, expected, actual)
+	})
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:139]: Would be cool to add one more, more sophisticated scenario. If you don't plan that, leave a `TODO` or something.

**Review by amwolff — COMMENTED** [review]
It's good, but I suggest to make some changes to make it even better.

File: internal/watcher/watcher_test.go (hunk 4 above)
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:139]: I'm planning to add it tomorrow evening.

File: internal/watcher/watcher_test.go (hunk 3 above)
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:107]: I settled upon a different solution for now.

File: internal/watcher/watcher_test.go (hunk 2 above)
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:76]: Ok

File: internal/watcher/watcher_test.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetAllInDirectory(t *testing.T) {
+	expectedDirCount := 2
+	expectedFilesCount := 0
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:16]: Ok

File: internal/watcher/watcher_test.go (hunk 4 above)
Comment (amwolff) [reply internal/watcher/watcher_test.go:139]: Ack. Hit the *Ready for review* whenever you're ready.

File: internal/watcher/watcher.go [hunk 6]
```diff
@@ -17,15 +17,21 @@ import (
...
 	walkFn := func(path string, info os.FileInfo, err error) error {
 		if err != nil {
 			return err
 		}
 
+		for _, e := range ignoredExtensions {
+			if filepath.Ext(path) == e {
+				return nil
+			}
+		}
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher.go:33]: Well, I don't know why I got rid of that in the first place.

File: internal/watcher/watcher.go [hunk 7]
```diff
@@ -17,15 +17,21 @@ import (
 )
 
 // GetAllInDirectory recursively returns all paths to files and directories in
-// dir. It returns nil, nil, <err> on the first error encountered.
-func GetAllInDirectory(dir string) ([]string, []string, error) {
+// dir(excluding files with ignored extensions). It returns nil, nil, <err> on the first error encountered.
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher.go:20]: Adjusted.

File: internal/watcher/watcher_test.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+		for {
+			<-time.After(100 * time.Millisecond)
+
+			if err := ioutil.WriteFile(files[0], []byte(":octopus: - :octopus:"), 0755); err != nil {
+				t.Logf("ioutil.WriteFile: %v", err)
+			}
+
+			if mockWatchFuncProvider.called {
+				break
+			}
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:126]: It's better to move this conditional to the beginning of the loop. It's a small improvement but may result in fewer writes.

File: internal/watcher/watcher_test.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+		previousContents := make(map[string][]byte)
+
+		baseDir, err := ioutil.TempDir("", "watcher_test1")
+		if err != nil {
+			t.Fatalf("ioutil.TempDir: %v", err)
+		}
+
+		defer os.RemoveAll(baseDir)
+
+		var b []byte
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:168]: You can assign this in the loop. It doesn't have to be in this scope, I guess.

**Review by amwolff — CHANGES_REQUESTED** [review]
Good job, the mutex-related comment needs attention. Apart from that, it's good to go.

File: internal/watcher/watcher_test.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+	m.calledMutex.Lock()
+	m.called = true
+	m.calledMutex.Unlock()
+
+	return nil
+}
+
+func TestWatchDirectories(t *testing.T) {
+	mockWatchFuncProvider := &mockWatchFuncProvider{
+		calledMutex: &sync.Mutex{},
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:94]: Ok

File: internal/watcher/watcher_test.go [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	done := make(chan struct{})
+
+	go func() {
+		for {
+			<-time.After(100 * time.Millisecond)
+
+			if err := ioutil.WriteFile(files[0], []byte(":octopus: - :octopus:"), 0755); err != nil {
+				t.Logf("ioutil.WriteFile: %v", err)
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:121]: Sure

File: internal/watcher/watcher_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:126]: Right

File: internal/watcher/watcher_test.go [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+func (m *mockWatchFuncProvider) WatchFn(path string) error {
+	m.calledMutex.Lock()
+	m.called = true
+	m.calledMutex.Unlock()
+
+	return nil
+}
+
+func TestWatchDirectories(t *testing.T) {
+	mockWatchFuncProvider := &mockWatchFuncProvider{
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:93]: Mock is definitely better

File: internal/watcher/watcher_test.go [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+				assert.Fail(t, "Haven't excluded files with ignored extensions.")
+				return
+			}
+		}
+	})
+}
+
+type mockWatchFuncProvider struct {
+	called      bool
+	calledMutex *sync.Mutex
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:81]: Ok

File: internal/watcher/watcher_test.go [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+		file = filepath.Join(subDir, fmt.Sprintf("test_file_%d", i))
+		if err = ioutil.WriteFile(file, []byte(":octopus:"), 0755); err != nil {
+			t.Fatalf("ioutil.WriteFile: %v", err)
+		}
+
+		file = filepath.Join(subDir, fmt.Sprintf("test_file_%d.swp", i))
+		if err = ioutil.WriteFile(file, []byte(":octopus:"), 0755); err != nil {
+			t.Fatalf("ioutil.WriteFile: %v", err)
+		}
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:53]: Done

File: internal/watcher/watcher.go [hunk 15]
```diff
@@ -16,16 +16,33 @@ import (
...
+
 // GetAllInDirectory recursively returns all paths to files and directories in
-// dir. It returns nil, nil, <err> on the first error encountered.
-func GetAllInDirectory(dir string) ([]string, []string, error) {
+// dir (excluding files with ignored extensions). It returns nil, nil, <err> on
+// the first error encountered.
+func GetAllInDirectory(dir string, ignoredExtensions []string) ([]string, []string, error) {
 	var files, directories []string
 
+	ignoredExtensionsMap := getIgnoredExtensions(ignoredExtensions)
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher.go:35]: Ok

File: internal/watcher/watcher_test.go (hunk 8 above)
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:126]: Nice tip

File: internal/watcher/watcher_test.go (hunk 9 above)
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:168]: My mistake

File: internal/watcher/watcher_test.go [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+	files, directories, err := GetAllInDirectory(baseDir, []string{".swp"})
+	if err != nil {
+		t.Fatalf("GetAllInDirectory: %v", err)
+	}
+
+	done := make(chan struct{})
+
+	go func() {
+		for {
+			<-time.After(100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go:118]: I'm not sure how it's supposed to work.

[the remark this answers is not in the document]
Comment (amwolff) [reply internal/watcher/watcher_test.go]: Ignore me. This won't work.

File: internal/watcher/watcher_test.go [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetAllInDirectory(t *testing.T) {
+	const (
+		filesMultiplier    = 10
+		expectedDirCount   = 2
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:20]:
```suggestion
		expectedDirsCount  = 2
```

File: internal/watcher/watcher_test.go [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+			file := filepath.Join(baseDir, fmt.Sprintf("test_file_%d", i))
+			if err = ioutil.WriteFile(file, []byte(fmt.Sprintf(":octopus:%d", i)), 0755); err != nil {
+				t.Fatalf("ioutil.WriteFile: %v", err)
+			}
+
+			var b []byte
+			b, err = ioutil.ReadFile(file)
+			if err != nil {
+				t.Fatalf("ioutil.ReadFile: %v", err)
+			}
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:185]:
```suggestion
			b, err := ioutil.ReadFile(file)
			if err != nil {
				t.Fatalf("ioutil.ReadFile: %v", err)
			}
```

**Review by amwolff — APPROVED** [review]

File: internal/watcher/watcher_test.go (hunk 17 above)
Comment (amwolff) [reply internal/watcher/watcher_test.go:20]: (this is not really important, so it's approved and up to you to apply).

File: internal/watcher/watcher_test.go (hunk 18 above)
Comment (amwolff) [reply internal/watcher/watcher_test.go:185]: (this is not really important, so it's approved and up to you to apply).

Comment (amwolff) [reply internal/watcher/watcher_test.go:185]: (this is not really important, so it's approved and up to you to apply).

Comment (DanielSz50) [reply internal/watcher/watcher_test.go:185]: When done this way, govet reports a shadow declaration of err.

File: internal/watcher/watcher_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (DanielSz50) [reply internal/watcher/watcher_test.go]: Ok

File: internal/watcher/watcher_test.go (hunk 17 above)
Comment (DanielSz50) [reply internal/watcher/watcher_test.go]: My mistake

**Review by amwolff — APPROVED** [review]

**Pull request merged**